src/l1_fetcher.rs

```diff
@@ -0,0 +1,123 @@
+use ethers::{abi::Contract, prelude::*, providers::Provider};
+use eyre::Result;
+use state_reconstruct::parse_calldata;
+use state_reconstruct::CommitBlockInfoV1;
+use state_reconstruct::BLOCK_STEP;
+use state_reconstruct::GENESIS_BLOCK;
+use state_reconstruct::ZK_SYNC_ADDR;
+use std::sync::Arc;
+use tokio::sync::mpsc;
+
+pub struct L1Fetcher {
+    provider: Provider<Http>,
+    contract: Contract,
+}
+
+impl L1Fetcher {
+    pub fn new(http_url: &str) -> Result<Self> {
+        let provider =
+            Provider::<Http>::try_from(http_url).expect("could not instantiate HTTP Provider");
+
+        let abi_file = std::fs::File::open("./IZkSync.json")?;
+        let contract = Contract::load(abi_file)?;
+
+        Ok(L1Fetcher { provider, contract })
+    }
+
+    pub async fn fetch(
+        &self,
+        sink: mpsc::Sender<Arc<Vec<CommitBlockInfoV1>>>,
+        start_block: Option<U64>,
+        end_block: Option<U64>,
+    ) -> Result<()> {
+        let mut current_l1_block_number = start_block.unwrap_or(GENESIS_BLOCK.into());
+        let end_block_number = end_block.unwrap_or(
+            self.provider
+                .get_block(BlockNumber::Latest)
+                .await?
+                .unwrap()
+                .number
+                .unwrap(),
+        );
+
+        let event = self.contract.events_by_name("BlockCommit")?[0].clone();
+        let function = self.contract.functions_by_name("commitBlocks")?[0].clone();
+
+        let (hash_tx, mut hash_rx) = mpsc::channel(5);
+        let (calldata_tx, mut calldata_rx) = mpsc::channel(5);
+
+        // Split L1 block processing into three async steps:
+        // - BlockCommit event filter.
+        // - Referred L1 block fetch.
+        // - Calldata parsing.
+        let provider = self.provider.clone();
+        tokio::spawn(async move {
+            while let Some(hash) = hash_rx.recv().await {
+                let tx = match provider.get_transaction(hash).await {
+                    Ok(x) => x,
+                    Err(e) => {
+                        println!("failed to get transaction for hash {}: {}", hash, e);
+                        continue;
+                    }
+                };
```

Comment:
Instead of skipping the hash on a provider error, we should probably retry a given number of times and if that doesn't succeed stop the fetching loop, especially if we're reconstructing from L1. However if we're just fetching we could just store the hashes of those missing blocks and proceed as normal, so maybe we need some sort of state machine :smile:

Reply:
This was a good shout. I'm not actually sure if we can skip any blocks at all, to maintain the state tree consistency, but I added retry there for now as a first mitigation. We need to consider if we want to abort processing on permanent error here.